### PR TITLE
FEATURE: LLM based peroidical summary report

### PR DESCRIPTION
**Pull request by SamSaffron**
Introduce a Discourse Automation based periodical report. Depends on Discourse Automation. 

Report works best with very large context language models such as GPT-4-Turbo and Claude 2. 

- Introduces `final_insts` to generic llm format, for claude to work best it is better to guide the last assistant message (we should add this to other spots as well) 
- Adds GPT-4 turbo support to generic llm interface

### CODE REVIEW

File: lib/automation/report_runner.rb [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            tokenizer: @llm.tokenizer,
+          )
+        input = <<~INPUT
+          #{@instructions}
+
+          <context>
+          #{context}
+          </context>
+
+          #{@instructions}
```
Comment (romanrizzi) [inline lib/automation/report_runner.rb:92]: Looks like we're adding instructions twice.

File: lib/automation/report_runner.rb [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            max_posts: @sample_size,
+            tags: @tags,
+            category_ids: @category_ids,
+            prioritized_group_ids: prioritized_group_ids,
+            allow_secure_categories: @allow_secure_categories,
+            tokens_per_post: @tokens_per_post,
+            tokenizer: @llm.tokenizer,
+          )
+        input = <<~INPUT
+          #{@instructions}
```
Comment (romanrizzi) [inline lib/automation/report_runner.rb:86]: I wonder if this should be added to `insts` instead. My understanding is that guidelines should be part of the system message?

**Review by romanrizzi — COMMENTED** [review]
If possible, I would like to see the runner tested. It should be easy with `Lllm#with_prepared_responses`

File: lib/completions/dialects/claude.rb [hunk 3]
```diff
@@ -17,7 +17,9 @@ def translate(generic_prompt)
 
           claude_prompt << "#{generic_prompt[:post_insts]}\n" if generic_prompt[:post_insts]
 
-          claude_prompt << "Assistant:\n"
+          claude_prompt << "Assistant:"
+          claude_prompt << " #{generic_prompt[:final_insts]}:" if generic_prompt[:final_insts]
```
Comment (xfalcox) [inline lib/completions/dialects/claude.rb:21]: Why are final instructions here under Assistant? Should the lines be swapped in order?

Comment (romanrizzi) [reply lib/completions/dialects/claude.rb:21]: This is what we talked the other day - Injecting the first sentence of the assistant response gives you a  free boost

File: lib/automation/report_runner.rb (hunk 1 above)
Comment (romanrizzi) [reply lib/automation/report_runner.rb:92]: TIL

File: lib/completions/dialects/claude.rb (hunk 3 above)
Comment (xfalcox) [reply lib/completions/dialects/claude.rb:21]: Ohhh it works well with Claude? That's awesome!

File: lib/automation/report_runner.rb (hunk 2 above)
Comment (SamSaffron) [reply lib/automation/report_runner.rb:86]: I don't think it helps, I tried it but this was the best way to ground it for giant contexts

File: lib/completions/dialects/claude.rb (hunk 3 above)
Comment (SamSaffron) [reply lib/completions/dialects/claude.rb:21]: yeah works super well we should add it everywhere ... saves us from so many annoyances

Comment (xfalcox) [reply lib/completions/dialects/claude.rb:31]: Yup, and if you are using XML tags to delimit the output, you can even put a small blurb + the opening tag to ensure a consistent reply from the bot.

**Review by romanrizzi — APPROVED** [review]

**Pull request merged**